eth/handler.go

```diff
@@ -288,22 +292,74 @@ func newHandler(config *handlerConfig) (*handler, error) {
 	return h, nil
 }
 
+// protoTracker tracks the number of active protocol handlers.
+func (h *handler) protoTracker() {
+	defer h.wg.Done()
+	var (
+		wg       sync.WaitGroup
+		done     = make(chan struct{})
+		active   int
+		quitting bool
+	)
+	// Start goroutine to track active protocol handlers.
+	wg.Add(1)
+	go func() {
+		for {
+			select {
+			case <-h.handlerStartCh:
+				active++
+			case <-h.handlerDoneCh:
+				active--
+			case <-done:
+				quitting = true
+			}
+			// Only check if active is 0 after quitSync has been
+			// closed. This ensures that in case active drops to 0
+			// naturally, this tracker routine stays operational.
+			if quitting && active == 0 {
+				wg.Done()
+				return
+			}
```

Comment:
One thing which I think is a problem is that even after quitting has been signalled, you still permit new handlers to be added. IMO you should do something like this
```golang
			select {
			case <-h.handlerStartCh:
				active++
			case <-h.handlerDoneCh:
				active--
			case <-h.quitSync:
				quitting = true
			}
			if quitting { 
				for active > 0 {
					<- handlerDoneCh
					active--
				}
				wg.Done()
				return
		        }
```
(maybe inline the quitting-thing and skip the `quitting` bool)

Reply:
Oh, and if you do this, you can skip a layer of `go func`. So no need to spin off a new goroutine, if you use `quitSync` instead of the dance with the `done`. But maybe I'm missing something there (?)